plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2536,9 +2662,184 @@ void producer_plugin::log_failed_transaction(const transaction_id_type& trx_id,
+      if ( exhausted ) {
+         lck.lock();
+         _ro_trx_queue.queue.push_front(trx);
+         break;
+      }
+   }
+
+   // If all tasks are finished, do not wait until end of read window; switch to write window now.
+   if ( --_ro_num_active_trx_exec_tasks == 0 ) {
+      // do switching on app thread
+      app().executor().post( priority::high, exec_queue::read_only_trx_safe, [self=this]() {
+         self->_ro_trx_exec_tasks_fut.clear();
+         self->switch_to_write_window();
+      } );
+   }
+
+   return true;
+}
+
+// Called from a read only trx thread. Run in parallel with app and other read only trx threads
+// Return exhausted status of the transaction.
+bool producer_plugin_impl::process_read_only_transaction(const packed_transaction_ptr& trx, const next_function<transaction_trace_ptr>& next, const fc::time_point& read_window_start_time) {
+   chain::controller& chain = chain_plug->chain();
+   auto future = transaction_metadata::start_recover_keys( trx, _thread_pool.get_executor(),
+                                                           chain.get_chain_id(), fc::microseconds::maximum(),
+                                                           transaction_metadata::trx_type::read_only,
+                                                           chain.configured_subjective_signature_length_limit() );
+   auto exception_handler = [&next](fc::exception_ptr ex) {
+      next( std::move(ex) );
+      return false;
+   };
+   future.wait();
+   try {
+      auto trx_metadata = future.get();
+      return push_read_only_transaction( trx_metadata, next, read_window_start_time );
+   } CATCH_AND_CALL(exception_handler);
+   return false;
+}
+
+// called on read_only_trx execution thread; multi-threaded safe 
+// return exhausted status
+bool producer_plugin_impl::push_read_only_transaction(
+            const transaction_metadata_ptr& trx,
+            next_function<transaction_trace_ptr> next,
+            const fc::time_point& read_window_start_time) {
+   auto exhausted = false;
+   chain::controller& chain = chain_plug->chain();
+
+   if( !chain.is_building_block() ) {
+      // try next round
+      return true;
+   }
+
+   try {
+      const auto block_deadline = calculate_block_deadline( chain.pending_block_time() );
+      auto start = fc::time_point::now();
+      auto time_used_in_read_window_us = ( start - read_window_start_time );
+      if ( time_used_in_read_window_us >= _ro_read_window_time_us ) {
+         // already passed read-window deadline, try next round
+         return true;
+      }
+
+      auto remained_time_in_read_window_us = _ro_read_window_time_us - time_used_in_read_window_us;
+      // set the trx to finish by the end of read-window. limit read window overrun
+      auto available_trx_time_us = std::min( remained_time_in_read_window_us, _ro_max_trx_time_us );
+
+      auto trace = chain.push_transaction( trx, block_deadline, available_trx_time_us, 0, false, 0 );
```

Comment:
The other call of `chain.push_transaction` still has:
```
   // setting and unsetting chainbase read_only mode will be moved to right
   // place when multi-threaded read-only transaction is implemented
   auto unset_db_read_only_mode = fc::make_scoped_exit([trx, &chain]{
      if( trx->is_read_only() )
         chain.unset_db_read_only_mode();
   });
   if( trx->is_read_only() )
      chain.set_db_read_only_mode();
```

Reply:
Thanks. Turns out the code needs to be stayed here but the comments are wrong. Correct the comments.